cdap-app-templates/cdap-etl/cdap-etl-api-spark/src/main/java/io/cdap/cdap/etl/api/batch/SparkCompute.java

```diff
@@ -70,13 +93,38 @@ public void initialize(SparkExecutionPluginContext context) throws Exception {
   }
 
   /**
-   * Transform the input and return the output to be sent to the next stage in the pipeline.
+   * Tells platform which version of transform call to use. Platform will generally follow the
+   * preference, but may still call RDD version even if preference is DataFrame. Platform will never
+   * call DataFrame version if it's not supported.
+   *
+   * @param context context {@link SparkExecutionPluginContext} for this job
+   * @return mode to use
+   */
+  public Mode getMode(SparkExecutionPluginContext context)
+      throws Exception {
+    return Mode.REQUIRES_RDD;
+  }
+
+  /**
+   * Transform the input RDD and return the output RDD to be sent to the next stage in the pipeline.
    *
    * @param context {@link SparkExecutionPluginContext} for this job
    * @param input input data to be transformed
    * @throws Exception if there is an error during this method invocation
    */
-  public abstract JavaRDD<OUT> transform(SparkExecutionPluginContext context, JavaRDD<IN> input)
+  public abstract JavaRDD<O> transform(SparkExecutionPluginContext context, JavaRDD<I> input)
```

Comment:
there may be cases where somebody only wants to implement the Dataset method, or where an equivalent RDD operation is not possible.

though I guess this could be done later by introducing another enum value for REQUIRES_DATAFRAME

Reply:
Well, as of now it's not always possible to create a DataFrame (e.g. schemaless run). Platform will try to use DataFrame if it's preferred, but can't guarantee it.

Reply:
But yes, we can add a mode in future that would tell platform to error out in validation if DataFrame mode is required but not available